sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuExpressions.scala

```diff
@@ -135,31 +174,28 @@ trait CudfUnaryExpression extends GpuUnaryExpression {
 trait GpuBinaryExpression extends BinaryExpression with GpuExpression {
 
   def doColumnar(lhs: GpuColumnVector, rhs: GpuColumnVector): ColumnVector
-  def doColumnar(lhs: Scalar, rhs: GpuColumnVector): ColumnVector
-  def doColumnar(lhs: GpuColumnVector, rhs: Scalar): ColumnVector
-  def doColumnar(numRows: Int, lhs: Scalar, rhs: Scalar): ColumnVector
+  def doColumnar(lhs: GpuScalar, rhs: GpuColumnVector): ColumnVector
+  def doColumnar(lhs: GpuColumnVector, rhs: GpuScalar): ColumnVector
+  def doColumnar(numRows: Int, lhs: GpuScalar, rhs: GpuScalar): ColumnVector
 
   override def columnarEval(batch: ColumnarBatch): Any = {
     withResourceIfAllowed(left.columnarEval(batch)) { lhs =>
       withResourceIfAllowed(right.columnarEval(batch)) { rhs =>
         (lhs, rhs) match {
           case (l: GpuColumnVector, r: GpuColumnVector) =>
             GpuColumnVector.from(doColumnar(l, r), dataType)
-          case (l, r: GpuColumnVector) =>
-            withResource(GpuScalar.from(l, left.dataType)) { scalar =>
-              GpuColumnVector.from(doColumnar(scalar, r), dataType)
-            }
-          case (l: GpuColumnVector, r) =>
-            withResource(GpuScalar.from(r, right.dataType)) { scalar =>
-              GpuColumnVector.from(doColumnar(l, scalar), dataType)
-            }
-          case (l, r) if l != null && r != null =>
-            withResource(GpuScalar.from(l, left.dataType)) { leftScalar =>
-              withResource(GpuScalar.from(r, right.dataType)) { rightScalar =>
-                GpuColumnVector.from(doColumnar(batch.numRows(), leftScalar, rightScalar), dataType)
-              }
-            }
-          case _ => null
+          case (l: GpuScalar, r: GpuColumnVector) =>
+            GpuColumnVector.from(doColumnar(l, r), dataType)
+          case (l: GpuColumnVector, r: GpuScalar) =>
+            GpuColumnVector.from(doColumnar(l, r), dataType)
+          case (l: GpuScalar, r: GpuScalar) =>
+            GpuColumnVector.from(doColumnar(batch.numRows(), l, r), dataType)
+          // null is not welcome, return a null Column instead.
+          case (l, r) if l == null || r == null =>
+            GpuColumnVector.fromNull(batch.numRows(), dataType)
```

Comment:
I am a little confused by this, because it appears that we have baked in null safe eval for scalars here, which is not the default for all binary expressions (like null safe equals). Why exactly do we need this? What situation do we run into where this path was taken?

Reply:
This existed before in both of the `GpuBinaryExpression ` and `GpuTernaryExpression`. Just keep it here since I do not know the history/background.

Reply:
I updated  #2412 to be a bug in the binary/ternary null handling. I am fine with that waiting because it did exist before this, but...

My main problem really was why is this looking for a null at all and not a GpuScalar that is wrapping a null.  Before this patch returning a null from a GpuExpression was returning a scalar null value. I thought the entire point of this code was to switch all of the scalar values over to return GpuScalar instances. If this patch was done properly then the null check above should be dead code and we can just remove it. But the null check is consistently in multiple places and with a comment on line 57 of GpuExpressions saying

> Null is legitimate in case of exceptions.

So my question was, what exceptions are those and why are they not an exception, but instead should return a null? It feels like we are masking bugs and I really want to understand it before I let this go in.

Reply:
I just removed that code and ran through all of the tests in the repo.  They all pass.  Could you just remove the null checks and let it be an error?  Then close #2412 when this is merged in?

```diff
diff --git a/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuExpressions.scala b/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuExpressions.scala
index 247b26424..7ac0cd9f0 100644
--- a/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuExpressions.scala
+++ b/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuExpressions.scala
@@ -53,8 +53,7 @@ object GpuExpressionsUtils extends Arm {
    * This is a common handling of the result from the `columnarEval`, allowing one of
    *   - A GpuColumnVector
    *   - A GpuScalar
-   *   - Null
-   * For other types, it will blow up. Null is legitimate in case of exceptions.
+   * For other types, it will blow up.
    *
    * It is recommended to return only a `GpuScalar` or a `GpuColumnVector` from a GPU
    * expression's `columnarEval`, to keep the result handling simple. Besides, `GpuScalar` can
@@ -71,7 +70,6 @@ object GpuExpressionsUtils extends Arm {
     withResourceIfAllowed(any) {
       case c: GpuColumnVector => c.incRefCount()
       case s: GpuScalar => GpuColumnVector.from(s, numRows, dType)
-      case null => GpuColumnVector.fromNull(numRows, dType)
       case other =>
         throw new IllegalArgumentException(s"Cannot resolve a ColumnVector from the value:" +
           s" $other. Please convert it to a GpuScalar or a GpuColumnVector before returning.")
@@ -190,9 +188,6 @@ trait GpuBinaryExpression extends BinaryExpression with GpuExpression {
             GpuColumnVector.from(doColumnar(l, r), dataType)
           case (l: GpuScalar, r: GpuScalar) =>
             GpuColumnVector.from(doColumnar(batch.numRows(), l, r), dataType)
-          // null is not welcome, return a null Column instead.
-          case (l, r) if l == null || r == null =>
-            GpuColumnVector.fromNull(batch.numRows(), dataType)
           case (l, r) =>
             throw new UnsupportedOperationException(s"Unsupported data '($l: " +
               s"${l.getClass}, $r: ${r.getClass})' for GPU binary expression.")
@@ -327,9 +322,6 @@ trait GpuTernaryExpression extends TernaryExpression with GpuExpression {
               GpuColumnVector.from(doColumnar(v0, v1, v2), dataType)
             case (v0: GpuScalar, v1: GpuScalar, v2: GpuScalar) =>
               GpuColumnVector.from(doColumnar(batch.numRows(), v0, v1, v2), dataType)
-            // null is not welcome, return a null column instead.
-            case (v0, v1, v2) if v0 == null || v1 == null || v2 == null =>
-              GpuColumnVector.fromNull(batch.numRows(), dataType)
             case (v0, v1, v2) =>
               throw new UnsupportedOperationException(s"Unsupported data '($v0: ${v0.getClass}," +
                 s" $v1: ${v1.getClass}, $v2: ${v2.getClass})' for GPU ternary expression.")

```

Reply:
Good to know and thanks for the verification. Updated.
I thought one case that the null was returned is for some exception handling, just like some expressions do under the non-ansi mode. e.g. GetMapValue returns null when key is not found. Now I think I was wrong, For GPU expressions, a null scalar or a null column vector suppose to be returned for this case.